### PR TITLE
Dependency resolver documentation revisions.

### LINKED ISSUE #3248: Document conda caching

### PR DESCRIPTION
**Pull request by jmchilton**
- Add documentation for cached dependency management as solution to LOCKERRORS.
- Add documentation for overcoming problems with hard links. (Different approaches)
- Add documentation on upgrading Conda as needed.
- Correct and fill out documentation of defaults.
- Fix some bugs in the Galaxy packages documentation.
- Expand Galaxy package documentation to include a simpler example.
- Update Conda configuration option documentation to reflect new defaults.
- Update Conda configuration option documentation to discuss copy_dependencies.
- Cross link Conda and dependency resolver documentation as it makes sense.
- Add a link to the Bioconda channel order documentation for future reference.
- Make things a bit more concise in places - start with the implicit tl;dr.

Fixes #3248

### CODE REVIEW

**Comment by jmchilton** [comment]
This does not yet cover mapping per se - but it is hopefully a very solid step forward.

File: doc/source/admin/conda_faq.rst [hunk 1]
```diff
@@ -25,8 +21,8 @@ community contributions (such as implementing a Guix package manager or
 enhancing the existing brew support to bring it on par with Conda).
 
 As a community, we have decided that Conda is the one that best fulfills
-community's needs. The following are some of the crucial Conda features that led
-to this decision:
+tje community's current needs. The following are some of the crucial Conda
```
Comment (mvdbeek) [inline doc/source/admin/conda_faq.rst]: s/tje/the

File: doc/source/admin/conda_faq.rst [hunk 2]
```diff
@@ -214,6 +203,11 @@ The order in which resolvers are tried is listed in the
 The first system that satisfies a requirement will be used. See
 `resolver docs`_ for detailed documentation.
 
+This however is not recommended, ideally tools will target and test
+against Conda for all dependencies. Also resolving all requirements
+with Conda gives Conda a chance to select compatibile versions of
```
Comment (mvdbeek) [inline doc/source/admin/conda_faq.rst]: s/compatibile/compatible

File: doc/source/admin/conda_faq.rst [hunk 3]
```diff
@@ -301,35 +295,110 @@ leave the old versions as they are – simply because of time.
...
+       The lock ['/galaxy/galaxy-app/tool-dependencies/_conda/pkgs/.conda_lock-119903'] was found. Wait for it to finish before continuing.
+       If you are sure that conda is not running, remove it and try again.
+       You can also use: $ conda clean --lock
+
+First, you may wish to enable cached dependencies. This can be done by setting
+``use_cached_dependency_manager`` in ``galaxy.ini``. Many jobs will create a
+per job Conda environment with just the dependencies needed for that job installed.
+This will be placed on the filesystem containg the job working directory. This
+is an expensive operation and Conda doesn't always link environments correctly
+accross filesystems. Enabling this job caching will create a cache for each required
```
Comment (mvdbeek) [inline doc/source/admin/conda_faq.rst]: s/accross/across/

File: doc/source/admin/conda_faq.rst [hunk 4]
```diff
@@ -301,35 +295,110 @@ leave the old versions as they are – simply because of time.
...
+as needed if the caching is in fact enabled (see `Pull Request #3348`_) and reduced
+the number of jobs that would require such caching (see `Pull Request #3391`_).
+
+
+15. What can I do about linking errors?
+***************************************
+
+If Galaxy jobs run on filesystems that cannot hardlink Conda packages managed
+by Galaxy, linking errors may occur when building environment to execute jobs.
+There are a few ways to potentially work around this discussed below.
```
Comment (mvdbeek) [inline doc/source/admin/conda_faq.rst]: s/ discussed below/

File: doc/source/admin/dependency_resolvers.rst [hunk 5]
```diff
@@ -33,32 +34,50 @@ The default configuration of dependency resolvers is equivalent to the following
...
+The default configuration thus prefers packages installed from the Galaxy Tool Shed using legacy ``tool_dependencies.xml``
+files, before trying to find a "Galaxy package" satisfying the specific version the dependency requires before
+falling back to looking for a Galaxy package with merely the correct name, and then looking for Conda recipes with
+matching name and version, and finally just for a Conda package with the correct name. If any of the dependency
 resolvers succeeds a dependency resolution object is returned and no more resolvers are called. This dependency
 resolution object provides shell commands to prepend to the shell script that runs the tool.
 
+This order can be thought of as a descending order of deliberation. Tool Shed dependencies must be declared next to the
+tool by the tool author and must be selected for installation at tool installation time - this requires specific actions
+by both the tool author and the deployer who isntalled the tools. The dependency is therefore expected to highly craft
```
Comment (mvdbeek) [inline doc/source/admin/dependency_resolvers.rst]: s/isntalled/installed

**Review by martenson — APPROVED** [review]

File: doc/source/admin/conda_faq.rst [hunk 6]
```diff
@@ -48,14 +44,19 @@ Below we answer some common questions (collected by Lance Parsons):
 1. How do I enable Conda dependency resolution for Galaxy jobs?
```
Comment (nsoranzo) [inline doc/source/admin/conda_faq.rst:44]: s/jobs/tools/

File: doc/source/admin/conda_faq.rst [hunk 7]
```diff
@@ -48,14 +44,19 @@ Below we answer some common questions (collected by Lance Parsons):
...
 ***************************************************************
 
-Galaxy's dependency job resolution is managed via
-``dependency_resolvers_conf.xml`` configuration file. Most Galaxy administrators
-should be using Galaxy's default dependency resolvers configuration file
-( ``dependency_resolvers_conf.xml.sample`` ). With
+The short answer is that as of 17.01, Galaxy should install Conda the first
+time it starts up and be configured to use it by default.
+
+The long answer is that Galaxy's dependency job resolution is managed via
```
Comment (nsoranzo) [inline doc/source/admin/conda_faq.rst:50]: s/dependency job/tool dependency/

File: doc/source/admin/conda_faq.rst [hunk 8]
```diff
@@ -48,14 +44,19 @@ Below we answer some common questions (collected by Lance Parsons):
...
+file is discussed in detail in the :ref:`Dependency Resolvers <dependency_resolvers>`
+documentation. Most Galaxy administrators will be using Galaxy's default dependency
+resolvers configuration file (``config/dependency_resolvers_conf.xml.sample``). With
 release 16.04, Galaxy has enabled Conda dependency resolution by default when
-Conda was already installed on the system. Having Conda enabled in
-``dependency_resolvers_conf.xml`` means that Galaxy can look for job
-dependencies using the Conda system when it attempts to run tools.
+Conda was already installed on the system. As of 17.01, Galaxy will also install
+Conda as needed when starting up. Having Conda enabled in ``dependency_resolvers_conf.xml``
+means that Galaxy can look for job dependencies using the Conda system when it
```
Comment (nsoranzo) [inline doc/source/admin/conda_faq.rst:58]: s/job/tool/

File: doc/source/admin/conda_faq.rst [hunk 9]
```diff
@@ -48,14 +44,19 @@ Below we answer some common questions (collected by Lance Parsons):
...
+attempts to run tools.
```
Comment (nsoranzo) [inline doc/source/admin/conda_faq.rst:59]: s/tools/a job/

File: doc/source/admin/conda_faq.rst (hunk 3 above)
Comment (nsoranzo) [inline doc/source/admin/conda_faq.rst]: s/accross/across/

File: doc/source/admin/conda_faq.rst [hunk 10]
```diff
@@ -301,35 +295,110 @@ leave the old versions as they are – simply because of time.
...
 .. code-block:: bash
 
-   ERROR: placeholder '/home/ray/r_3_3_1-x64-3.5/envs/_build_placehold_placehold_placehold_placehold_pl' too short
+   Error:     LOCKERROR: It looks like conda is already doing something.
+       The lock ['/galaxy/galaxy-app/tool-dependencies/_conda/pkgs/.conda_lock-119903'] was found. Wait for it to finish before continuing.
+       If you are sure that conda is not running, remove it and try again.
+       You can also use: $ conda clean --lock
+
+First, you may wish to enable cached dependencies. This can be done by setting
+``use_cached_dependency_manager`` in ``galaxy.ini``. Many jobs will create a
```
Comment (nsoranzo) [inline doc/source/admin/conda_faq.rst:334]: Add: `Without this option,` before `Many jobs...`

File: doc/source/admin/conda_faq.rst [hunk 11]
```diff
@@ -301,35 +295,110 @@ leave the old versions as they are – simply because of time.
...
+lie on the same file system as the default Conda installation.
+
+If this still doesn't work, perhaps the underlying file system does not support hard
+linking at all. In this case it is best to add ``always_softlink: True`` to Galaxy's
+YAML ``condarc`` file, this should be created by Galaxy and placed in
+``<tool_dependency_dir/_condarc``. This requires Conda 4.3 or newer. Note this is a
+newer version of Conda than shipped with Galaxy as of 17.01. See the question below
+on upgrading Conda if you must use this trick.
+
+Alternatively, copying can be used when creating environments instead links (either
```
Comment (nsoranzo) [inline doc/source/admin/conda_faq.rst:367]: s/instead links/instead of links/

**Comment by jmchilton** [comment]
@martenson and @mvdbeek - thanks for cleaning up my mess and thanks for the quick merge!

**Comment by martenson** [comment]
@jmchilton thanks for the docs dude!

File: doc/source/admin/dependency_resolvers.rst [hunk 12]
```diff
@@ -163,32 +195,43 @@ and will almost certainy be removed from the code base.
...
 debug
-    Pass debug flag to conda commands (default: false).
+    Pass debug flag to conda commands (default: *False*).
 
 ensure_channels
     conda channels to enable by default. See
     http://conda.pydata.org/docs/custom-channels.html for more
-    information about channels. (default: iuc,bioconda,r,defaults,conda-forge).
+    information about channels. This defaults to ``iuc,bioconda,r,defaults,conda-forge``.
+    This order should be consistent with `Bioconda perscribed order <https://github.com/bioconda/bioconda-recipes/blob/master/config.yml#L8>`__
```
Comment (nsoranzo) [inline doc/source/admin/dependency_resolvers.rst:221]: s/perscribed/prescribed/

**Comment by nsoranzo** [comment]
This PR targeted the wrong branch!

**Pull request merged**